File: IdeaProjects/Tienda_py/.idea/Menu/Compra.py
from datetime import date
from datetime import datetime
from datetime import datetime
from Cliente import Cliente
from Productos.Producto import Producto
from Productos.Alimento import Alimento
from Productos.Belleza import Belleza
from Productos.Electrodomestico import Electrodomestico
from Productos.Limpieza import Limpieza

class Compra:
    def _init_(self):
        self.productos = []
        self.clientes = []
        self.cantidadP = 0
        self.cantidadC = 0

    def comprar(self, idC, idP, cantidad):
        precio = 0
        for producto in self.productos:
            if idP == producto.get_id():
                if producto.reducir_stock(cantidad):
                    precio = producto.get_precio() * cantidad
                    for cliente in self.clientes:
                        if idC == cliente.get_id():
                            cliente.registrar_producto(producto, cantidad)
                break
        return precio

    def add_producto(self, producto):
        producto.set_id(self.cantidadP)
        self.productos.append(producto)
        self.cantidadP += 1

    def aumentar_stock(self, id, cantidad):
        for producto in self.productos:
            if id == producto.get_id():
                producto.aumentar_stock(cantidad)
                break

    def modificar_producto(self, id):
        opcion = 0
        while opcion != 8:
            for producto in self.productos:
                if id == producto.get_id():
                    print("Ingrese la opción a modificar: ")
                    print("1 Nombre")
                    print("2 Tipo")
                    print("3 Precio")
                    print("4 Stock")
                    print("5 Fecha de importación")
                    print("6 Número de serie")
                    print("7 Otros")
                    print("8 Salir de la modificación")
                    opcion = int(input())
                    if opcion == 1:
                        name = input("Ingrese nuevo valor: ")
                        producto.set_name(name)
                    elif opcion == 2:
                        tipo = input("Ingrese nuevo valor: ")
                        producto.set_tipo(tipo)
                    elif opcion == 3:
                        precio = float(input("Ingrese nuevo valor: "))
                        producto.set_precio(precio)
                    elif opcion == 4:
                        stock = int(input("Ingrese nuevo valor: "))
                        producto.set_stock(stock)
                    elif opcion == 5:
                        fecha_importacion = self.asignar_fecha()
                        producto.set_fecha_importacion(fecha_importacion)
                    elif opcion == 6:
                        no = int(input("Ingrese nuevo valor: "))
                        producto.set_no(no)
                    elif opcion == 7:
                        if isinstance(producto, Alimento):
                            caracter = input("Cambiar fecha de caducidad (ingresar carácter, 0 para no): ")
                            if caracter != "0":
                                caducidad = self.asignar_fecha()
                                producto.set_fecha_importacion(caducidad)
                        elif isinstance(producto, Limpieza):
                            op = 0
                            while op != 3:
                                print("Ingrese la opción a modificar: ")
                                print("1 Aroma")
                                print("2 Recomendado por expertos")
                                print("3 Salir")
                                op = int(input())
                                if op == 1:
                                    aroma = input("Ingresar nuevo aroma: ")
                                    producto.set_aroma(aroma)
                                elif op == 2:
                                    caracter = input("¿El producto es recomendado por al menos 8 de cada 10 expertos? (0 para NO, otro carácter para SÍ): ")
                                    producto.set_recomend(caracter != "0")
                                else:
                                    print("Esa opción no se encuentra")
                        elif isinstance(producto, Electrodomestico):
                            op = 0
                            while op != 3:
                                print("Ingrese la opción a modificar: ")
                                print("1 Color")
                                print("2 Watts")
                                print("3 Descripción del producto")
                                print("4 Salir")
                                op = int(input())
                                if op == 1:
                                    color = input("Ingresar nuevo color: ")
                                    producto.set_color(color)
                                elif op == 2:
                                    watts = float(input("Ingresar watts: "))
                                    producto.set_watts(watts)
                                elif op == 3:
                                    descripcion = input("Ingresar descripción: ")
                                    producto.set_descripcion(descripcion)
                                else:
                                    print("Esa opción no se encuentra")
                        elif isinstance(producto, Belleza):
                            op = 0
                            while op != 3:
                                print("Ingrese la opción a modificar: ")
                                print("1 Color")
                                print("2 Duración diaria")
                                print("3 Salir")
                                op = int(input())
                                if op == 1:
                                    color = input("Ingresar nuevo color: ")
                                    producto.set_color(color)
                                elif op == 2:
                                    duracion = int(input("Ingresar duración diaria en horas enteras: "))
                                    producto.set_duracion_diaria(duracion)
                                else:
                                    print("Esa opción no se encuentra")

    def asignar_fecha(self):
        while True:
            try:
                fecha_str = input("Ingrese la fecha (AAAA-MM-DD): ")
                return datetime.strptime(fecha_str, "%Y-%m-%d").date()
            except ValueError:
                print("Formato de fecha incorrecto. Intente nuevamente.")

    def mostrar_productos():
        i = 0
        while i < 4:
            if i == 0:
                print("Alimentos: ")
            if i == 1:
                print("Limpieza: ")
            if i == 2:
                print("Electrodomesticos: ")
            if i == 3:
                print("Belleza: ")

            for producto in productos:
                if isinstance(producto, Alimento) and i == 0:
                    print(producto.get_name(), producto.get_tipo(), producto.get_id(), producto.get_stock())
                if isinstance(producto, Limpieza) and i == 1:
                    print(producto.get_name(), producto.get_tipo(), producto.get_id(), producto.get_stock())
                if isinstance(producto, Electrodomestico) and i == 2:
                    print(producto.get_name(), producto.get_tipo(), producto.get_id(), producto.get_stock())
                if isinstance(producto, Belleza) and i == 3:
                    print(producto.get_name(), producto.get_tipo(), producto.get_id(), producto.get_stock())

            i += 1

    def mostrar_producto(id):
        for producto in productos:
            if id == producto.get_id():
                print("Producto: " + str(producto))

    def eliminar_producto(id):
        for producto in productos:
            if id == producto.get_id():
                productos.remove(producto)

    def add_cliente(cliente):
        clientes.append(cliente)
        cliente.set_id(cantidadC)
        cantidadC += 1

    def modificar_cliente(id):
        opcion = 0
        while opcion != 3:
            for cliente in clientes:
                if id == cliente.get_id():
                    print("Ingrese la opción a modificar: ")
                    print("1 Nombre")
                    print("2 Edad")
                    print("3 Salir")
                    opcion = int(input())
                    if opcion == 1:
                        nombre = input("Ingrese el nombre: ")
                        cliente.set_nombre(nombre)
                    elif opcion == 2:
                        edad = int(input("Ingrese la edad: "))
                        cliente.set_edad(edad)
                    elif opcion == 3:
                        print("Saliendo del menú")
                    else:
                        print("Esa opción no se encuentra")

        def mostrar_clientes():
            for cliente in clientes:
                print(cliente)

        def mostrar_cliente(id):
            for cliente in clientes:
                if id == cliente.get_id():
                    print(cliente)

        def eliminar_cliente(id):
            for cliente in clientes:
                if id == cliente.get_id():
                    clientes.remove(cliente)

File: IdeaProjects/Tienda/src/Menu/Compra.java
package Menu;

import Productos.*;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Scanner;

public class Compra {
    Scanner leer = new Scanner(System.in);
    ArrayList<Producto> productos = new ArrayList<>();
    ArrayList<Cliente> clientes = new ArrayList<>();

    private int cantidadP, cantidadC;
    //Esta cantidad sigue contando a los clientes que han sido eliminados, pues solo sirve para no repetir el id


    public double comprar(int idC, int idP, int cantidad) {
        boolean cPudo;
        double precio = 0;
        for (Producto producto : productos) {
            if (idP == producto.getId()) {
                cPudo = producto.reducirStock(cantidad);
                if (cPudo) {
                    precio = producto.getPrecio() * cantidad;
                    for (Cliente cliente : clientes) {
                        if (idC == cliente.getId()) {
                            cliente.registrarProducto(producto, cantidad);
                        }
                    }
                }

            }
        }
        return precio;

    }


    public void addProducto(Producto producto) {
        producto.setId(cantidadP);
        productos.add(producto);
        cantidadP++;

    }
    public void aumentarStock(int id,int cantidad){
        for (Producto producto:productos){
            if(id==producto.getId()){
                producto.aumentarStock(cantidad);
            }

        }
    }

    public void modificarProducto(int id) {
        String name;
        double precio;
        int stock;
        int no;
        LocalDate fechaImportacion;
        String tipo;
        int opcion = 0;
        while (opcion != 8) {
            for (Producto producto : productos) {
                if (id == producto.getId()) {
                    System.out.println("Ingrese la opción a modificar: ");
                    System.out.println("1 Nombre");
                    System.out.println("2 Tipo");
                    System.out.println("3 Precio");
                    System.out.println("4 Stock");
                    System.out.println("5 Fecha de importación");
                    System.out.println("6 Número de serie");
                    System.out.println("7 Otros");
                    System.out.println("8 Salir de la modificación");
                    opcion = leer.nextInt();
                    switch (opcion) {
                        case 1:
                            System.out.println("Ingrese nuevo valor: ");
                            name = leer.nextLine();
                            producto.setName(name);
                            break;
                        case 2:
                            System.out.println("Ingrese nuevo valor: ");
                            tipo = leer.nextLine();
                            producto.setTipo(tipo);
                            break;
                        case 3:
                            System.out.println("Ingrese nuevo valor: ");
                            precio = leer.nextDouble();

                            leer.nextLine();
                            producto.setPrecio(precio);
                            break;
                        case 4:
                            System.out.println("Ingrese nuevo valor: ");
                            stock = leer.nextInt();

                            leer.nextLine();
                            producto.setStock(stock);
                            break;
                        case 5:
                            fechaImportacion = asignarFecha();
                            producto.setFechaImportacion(fechaImportacion);

                            break;
                        case 6:
                            System.out.println("Ingrese nuevo valor: ");
                            no = leer.nextInt();

                            leer.nextLine();
                            producto.setNo(no);
                            break;
                        case 7:
                            if (producto instanceof Alimento) {
                                String caracter;
                                System.out.println("Cambiar fecha de caducidad: ingresar carácter, sí no lo desea ingrese 0");
                                caracter = leer.nextLine();
                                if (!caracter.equals("0")) {
                                    LocalDate caducidad = asignarFecha();
                                    producto.setFechaImportacion(caducidad);
                                }
                            }
                            if (producto instanceof Limpieza) {
                                int op = 0;
                                while (op != 3) {
                                    System.out.println("Ingrese la opción a modificar: ");
                                    System.out.println("1 Aroma");
                                    System.out.println("2 Recomendado por expertos");
                                    System.out.println("3 Salir");
                                    op = leer.nextInt();

                                    leer.nextLine();
                                    switch (op) {
                                        case 1:
                                            System.out.println("Ingresar nuevo aroma: ");
                                            String aroma = leer.nextLine();
                                            break;
                                        case 2:
                                            System.out.println("¿El producto es recomendado por al menos 8 de cada 10 expertos?: ");
                                            System.out.println("Digite 0 sí la respuesta es NO, digite otro carácte sí la respuesta es si");
                                            String caracter = leer.nextLine();
                                            if (!caracter.equals("0")) {
                                                ((Limpieza) producto).setRecomend(true);
                                            } else {
                                                ((Limpieza) producto).setRecomend(false);
                                            }
                                            break;
                                        default:
                                            System.out.println("Esa opción no se encuentra");
                                            break;
                                    }
                                }

                            }
                            if (producto instanceof Electrodomestico) {

                                int op = 0;
                                while (op != 3) {
                                    System.out.println("Ingrese la opción a modificar: ");
                                    System.out.println("1 Color");
                                    System.out.println("2 Watts");
                                    System.out.println("3 Descripción del producto");
                                    System.out.println("4 Salir");
                                    op = leer.nextInt();

                                    leer.nextLine();
                                    switch (op) {
                                        case 1:
                                            System.out.println("Ingresar nuevo color: ");
                                            String color = leer.nextLine();
                                            ((Electrodomestico) (producto)).setColor(color);
                                            break;
                                        case 2:
                                            System.out.println("Ingresar watts: ");
                                            double watts = leer.nextDouble();
                                            ((Electrodomestico) (producto)).setWatts(watts);

                                            break;
                                        case 3:
                                            System.out.println("Ingresar descripción: ");
                                            String descripcion = leer.nextLine();
                                            ((Electrodomestico) (producto)).setDescripcion(descripcion);

                                            break;
                                        default:
                                            System.out.println("Esa opción no se encuentra");
                                            break;
                                    }
                                }

                            }
                            if (producto instanceof Belleza) {
                                int op = 0;
                                while (op != 3) {
                                    System.out.println("Ingrese la opción a modificar: ");
                                    System.out.println("1 Color");
                                    System.out.println("2 Duración diaria");
                                    System.out.println("3 Salir");
                                    op = leer.nextInt();

                                    leer.nextLine();
                                    switch (op) {
                                        case 1:
                                            System.out.println("Ingresar nuevo color: ");
                                            String color = leer.nextLine();
                                            ((Belleza) producto).setColor(color);
                                            break;
                                        case 2:
                                            System.out.println("Ingresar duración diaria en horas enteras");
                                            int duracion = leer.nextInt();
                                            leer.nextLine();
                                            ((Belleza) producto).setDuracionDiaria(duracion);
                                            break;
                                        default:
                                            System.out.println("Esa opción no se encuentra");
                                            break;
                                    }
                                }

                            }
                            break;
                        case 8:
                            System.out.println("Usted ha salido de modificar");
                            break;
                        default:
                            System.out.println("Esa opción no se encuentra");
                            break;


                    }
                }
            }

        }
    }

    public LocalDate asignarFecha() {
        int dia = 43, mes = 1, ano = 2024;
        LocalDate fecha;
        while ((dia > 32 || dia < 0) || (mes < 0 || mes > 12) || (ano < 2023)) {
            System.out.println("Ingrese día: ");
            dia = leer.nextInt();

            leer.nextLine();
            System.out.println("Ingrese mes: ");
            mes = leer.nextInt();

            leer.nextLine();
            System.out.println("Ingrese año: ");
            ano = leer.nextInt();

            leer.nextLine();

        }
        fecha = LocalDate.of(ano, mes, dia);
        return fecha;
    }

    public void mostrarProductos() {
        int i = 0;
        while (i < 4) {
            if (i == 0) {
                System.out.println("Alimentos: ");
            }
            if (i == 1) {
                System.out.println("Limpieza: ");
            }
            if (i == 2) {
                System.out.println("Electrodomesticos: ");
            }
            if (i == 3) {
                System.out.println("Belleza: ");
            }

            for (Producto producto : productos) {

                if (producto instanceof Alimento && i == 0) {
                    System.out.println(producto.getName() + " " + producto.getTipo() + " " + producto.getId() + " " + producto.getStock());
                }
                if (producto instanceof Limpieza && i == 1) {
                    System.out.println(producto.getName() + " " + producto.getTipo() + " " + producto.getId() + " " + producto.getStock());
                }
                if (producto instanceof Electrodomestico && i == 2) {
                    System.out.println(producto.getName() + " " + producto.getTipo() + " " + producto.getId() + " " + producto.getStock());
                }
                if (producto instanceof Belleza && i == 3) {
                    System.out.println(producto.getName() + " " + producto.getTipo() + " " + producto.getId() + " " + producto.getStock());
                }
            }

            i++;
        }
    }


    public void mostrarProducto(int id) {
        for (Producto producto : productos) {
            if (id == producto.getId()) {
                System.out.println("Producto: " + (producto).toString());
            }
        }

    }

    public void eliminarProducto(int id) {
        for (Producto producto : productos) {
            if (id == producto.getId()) {
                productos.remove(producto);
            }
        }
    }

    public void addCliente(Cliente cliente) {
        clientes.add(cliente);
        cliente.setId(cantidadC);
        cantidadC++;

    }

    public void modificarCliente(int id) {
        String nombre;
        int edad;
        int opcion = 0;
        while (opcion != 3) {
            for (Cliente cliente : clientes) {
                if (id == cliente.getId()) {
                    System.out.println("Ingrese la opción a modificar: ");
                    System.out.println("1 Nombre");
                    System.out.println("2 Edad");
                    System.out.println("3 Salir");
                    opcion = leer.nextInt();
                    leer.nextLine();
                    switch (opcion) {
                        case 1:
                            System.out.println("Ingrese el nombre: ");
                            nombre=leer.nextLine();
                            cliente.setNombre(nombre);
                            break;
                        case 2:
                            System.out.println("Ingrese la edad: ");
                            edad=leer.nextInt();
                            leer.nextLine();
                            cliente.setEdad(edad);
                            break;
                        case 3:
                            System.out.println("Saliendo del menú");
                            break;
                        default:
                            System.out.println("Esa opción no se encuentra");
                            break;
                    }
                }
            }
        }
    }

    public void mostrarClientes() {
        for (Cliente cliente : clientes) {
            System.out.println(cliente.toString());
        }

    }

    public void mostrarCliente(int id) {
        for (Cliente cliente : clientes) {
            if (id == cliente.getId()) {
                System.out.println(cliente.toString());
            }
        }

    }

    public void eliminarCliente(int id) {
        for (Cliente cliente : clientes) {
            if (id == cliente.getId()) {
                clientes.remove(cliente);
            }
        }


    }
}
